### PR TITLE
chore(audoedit): simplify diff utils and renderer data structures

### PR DESCRIPTION
**Pull request by valerybugakov**
- No functional changes.
- Simplifies data structures used to calculate the diff between the current document content and predicted edits.
- Simplifies data structures used to render autoedit decorations. 
- Removes redundant data structures and transformations happening between the diff calculation and the renderer call. 
- Adds granular `changes: LineChange` to `ModifiedLineInfo`, representing individual insertions and deletions in modified lines. This is helpful for troubleshooting purposes, and I plan to use it for the experimental inline renderer implementation that we discussed recently in Slack.
- Updates the default autoedits renderer to use all the above changes.
- Built on top of https://github.com/sourcegraph/cody/pull/6163

## Test plan

Manually tested; updated existing diff-utils unit tests and added new ones to cover `LineChange` calculations.

### CODE REVIEW

File: vscode/src/autoedits/renderer/decorators/base.ts [hunk 1]
```diff
@@ -25,46 +24,56 @@ import type { ModifiedRange } from '../diff-utils'
...
 
-export interface DecorationInformation {
-    lines: DecorationLineInformation[]
-    oldLines: string[]
-    newLines: string[]
+export interface DecorationInfo {
+    modifiedLines: ModifiedLineInfo[]
+    removedLines: RemovedLineInfo[]
+    addedLines: AddedLineInfo[]
+    unchangedLines: UnchangedLineInfo[]
```
Comment (hitesh-1997) [inline vscode/src/autoedits/renderer/decorators/base.ts:78]: nice !! This contract is much better than what I had before.

**Comment by valerybugakov** [comment]
@hitesh-1997, I gave it my best shot with manual testing (which is how I found https://github.com/sourcegraph/cody/pull/6171), but this is still a pretty sensitive area without any tests yet (referring to the default renderer). Let me know if you spot anything during your manual testing!

File: vscode/src/autoedits/renderer/decorators/default-decorator.ts [hunk 2]
```diff
@@ -387,3 +354,15 @@ function getCommonPrefix(s1: string, s2: string): string {
     }
     return commonPrefix
 }
+
+/**
+ * Checks if the only changes for modified lines are additions of text.
+ */
+export function isOnlyAddingTextForModifiedLines(modifiedLines: ModifiedLineInfo[]): boolean {
```
Comment (hitesh-1997) [inline vscode/src/autoedits/renderer/decorators/default-decorator.ts:361]: I added this in the common assuming other decorators might need it too. But I think we can keep it here for now, and if other decorator need it we can move it to common.

Comment (valerybugakov) [reply vscode/src/autoedits/renderer/decorators/default-decorator.ts:361]: Yeah, I didn't mean it as an important change we must make. I tried to understand the code better and moved things around multiple times.

File: vscode/src/autoedits/renderer/decorators/base.ts [hunk 3]
```diff
@@ -25,46 +24,56 @@ import type { ModifiedRange } from '../diff-utils'
...
+
+export interface UnchangedLineInfo {
+    type: 'unchanged'
+    text: string
+    lineNumber: number // `lineNumber` in the modified text
+}
+
+export type LineChange = {
+    type: 'insert' | 'delete'
+    range: vscode.Range
```
Comment (hitesh-1997) [inline vscode/src/autoedits/renderer/decorators/base.ts:70]: Is this range the index relative to the start of the text or the start of the document. A comment would help clarify this.

File: vscode/src/autoedits/renderer/diff-utils.ts [hunk 4]
```diff
@@ -1,241 +1,261 @@
...
-    from1: number
-    /** The end position in the original text */
-    to1: number
-    /** The start position in the modified text */
-    from2: number
-    /** The end position in the modified text */
-    to2: number
+export function getDecorationInfo(originalText: string, modifiedText: string): DecorationInfo {
+    const originalLines = originalText.split('\n')
+    const modifiedLines = modifiedText.split('\n')
```
Comment (hitesh-1997) [inline vscode/src/autoedits/renderer/diff-utils.ts:15]: should we use lines function here to make it agnostic to windows.

Comment (valerybugakov) [reply vscode/src/autoedits/renderer/diff-utils.ts:15]: This is a great call. I have a couple of other places that should be migrated to platform-agnostic line separators. I will do that.

**Review by hitesh-1997 — APPROVED** [review]
Nice !!
This removes a bunch of duplications

**Comment by hitesh-1997** [comment]
> @hitesh-1997, I gave it my best shot with manual testing (which is how I found #6171), but this is still a pretty sensitive area without any tests yet (referring to the default renderer). Let me know if you spot anything during your manual testing!

This makes sense @valerybugakov 
I didn't find anything breaking with manual testing but since we are starting to make more frequent changes, we should definitely prioritize tests, will try to wrap up the diff logic PR from my end so that we can move to testing :)

**Pull request merged**